File: dotnet/src/Envdoctor/Runtime/Token.cs
using System.IO.Compression;
using System.Text;
using System.Text.Json;
using Envdoctor.Models;
using Envdoctor.Utils;

namespace Envdoctor.Runtime;

/// Portable snapshot token: `base64url(gzip(json))` with an `envd1:` prefix.
public static class Token
{
    private const string Prefix = "envd1:";

    /// Serialize a snapshot to compact JSON with the reference property order.
    public static string SnapshotToJson(RuntimeSnapshot snapshot)
    {
        var globals = new JsonObject();
        foreach (var (eco, packages) in snapshot.Globals)
        {
            globals.Add(eco, packages
                .Select(p => (object)new JsonObject { { "name", p.Name }, { "version", p.Version } })
                .ToList());
        }
        return Json.Compact(new JsonObject
        {
            { "schema", snapshot.Schema },
            { "capturedAt", snapshot.CapturedAt },
            {
                "os", new JsonObject
                {
                    { "platform", snapshot.Os.Platform },
                    { "arch", snapshot.Os.Arch },
                    { "release", snapshot.Os.Release },
                }
            },
            {
                "tools", snapshot.Tools
                    .Select(t => (object)new JsonObject
                    {
                        { "tool", t.Tool },
                        { "version", t.Version },
                        { "resolvedFrom", t.ResolvedFrom },
                    })
                    .ToList()
            },
            { "path", snapshot.Path.Cast<object?>().ToList() },
            { "globals", globals },
            { "envFlagNames", snapshot.EnvFlagNames.Cast<object?>().ToList() },
        });
    }

    /// Encode a snapshot into a single-line, paste-safe token.
    public static string EncodeToken(RuntimeSnapshot snapshot)
    {
        var json = Encoding.UTF8.GetBytes(SnapshotToJson(snapshot));
        using var output = new MemoryStream();
        using (var gzip = new GZipStream(output, CompressionLevel.Optimal, leaveOpen: true))
            gzip.Write(json, 0, json.Length);
        return Prefix + Base64UrlEncode(output.ToArray());
    }

    /// Decode a token back into a snapshot. Errors clearly on malformed or
    /// too-new input.
    public static RuntimeSnapshot DecodeToken(string token)
    {
        var trimmed = token.Trim();
        if (!trimmed.StartsWith(Prefix, StringComparison.Ordinal))
            throw new SnapshotTokenException("Not an envdoctor snapshot token (missing envd1: prefix).");

        RuntimeSnapshot snapshot;
        try
        {
            var gzipped = Base64UrlDecode(trimmed[Prefix.Length..]);
            using var input = new MemoryStream(gzipped);
            using var gzip = new GZipStream(input, CompressionMode.Decompress);
            using var reader = new StreamReader(gzip, Encoding.UTF8);
            snapshot = ParseSnapshot(JsonDocument.Parse(reader.ReadToEnd()).RootElement);
        }
        catch (Exception)
        {
            throw new SnapshotTokenException("Corrupt snapshot token: could not decode.");
        }
        AssertReadable(snapshot);
        return snapshot;
    }

    /// Parse raw JSON (from a `--output` file) into a validated snapshot.
    public static RuntimeSnapshot ParseSnapshotJson(string text)
    {
        RuntimeSnapshot snapshot;
        try
        {
            snapshot = ParseSnapshot(JsonDocument.Parse(text).RootElement);
        }
        catch (SnapshotTokenException)
        {
            throw;
        }
        catch (Exception)
        {
            throw new SnapshotTokenException("Invalid snapshot JSON.");
        }
        AssertReadable(snapshot);
        return snapshot;
    }

    /// Reject snapshots from a newer schema than this build understands.
    private static void AssertReadable(RuntimeSnapshot snapshot)
    {
        if (string.CompareOrdinal(snapshot.Schema, SnapshotSchema.Value) > 0)
        {
            throw new SnapshotTokenException(
                $"Snapshot schema {snapshot.Schema} is newer than this envdoctor ({SnapshotSchema.Value}). Upgrade to compare it.");
        }
    }

    private static RuntimeSnapshot ParseSnapshot(JsonElement el)
    {
        if (el.ValueKind != JsonValueKind.Object)
            throw new SnapshotTokenException("Invalid snapshot JSON.");

        var snapshot = new RuntimeSnapshot();
        if (el.TryGetProperty("schema", out var schema) && schema.ValueKind == JsonValueKind.String)
            snapshot.Schema = schema.GetString() ?? "";
        if (el.TryGetProperty("capturedAt", out var capturedAt) && capturedAt.ValueKind == JsonValueKind.String)
            snapshot.CapturedAt = capturedAt.GetString() ?? "";
        if (el.TryGetProperty("os", out var os) && os.ValueKind == JsonValueKind.Object)
        {
            snapshot.Os = new OsInfo
            {
                Platform = GetStr(os, "platform"),
                Arch = GetStr(os, "arch"),
                Release = GetStr(os, "release"),
            };
        }
        if (el.TryGetProperty("tools", out var tools) && tools.ValueKind == JsonValueKind.Array)
        {
            snapshot.Tools = tools.EnumerateArray()
                .Select(t => new ToolInfo
                {
                    Tool = GetStr(t, "tool"),
                    Version = GetStr(t, "version"),
                    ResolvedFrom = GetStr(t, "resolvedFrom"),
                })
                .ToList();
        }
        if (el.TryGetProperty("path", out var path) && path.ValueKind == JsonValueKind.Array)
            snapshot.Path = path.EnumerateArray().Select(p => p.GetString() ?? "").ToList();
        if (el.TryGetProperty("globals", out var globals) && globals.ValueKind == JsonValueKind.Object)
        {
            snapshot.Globals = globals.EnumerateObject().ToDictionary(
                eco => eco.Name,
                eco => eco.Value.ValueKind == JsonValueKind.Array
                    ? eco.Value.EnumerateArray()
                        .Select(p => new GlobalPackage { Name = GetStr(p, "name"), Version = GetStr(p, "version") })
                        .ToList()
                    : new List<GlobalPackage>());
        }
        if (el.TryGetProperty("envFlagNames", out var flags) && flags.ValueKind == JsonValueKind.Array)
            snapshot.EnvFlagNames = flags.EnumerateArray().Select(p => p.GetString() ?? "").ToList();
        return snapshot;
    }

    private static string GetStr(JsonElement el, string name) =>
        el.ValueKind == JsonValueKind.Object &&
        el.TryGetProperty(name, out var v) &&
        v.ValueKind == JsonValueKind.String
            ? v.GetString() ?? ""
            : "";

    private static string Base64UrlEncode(byte[] data) =>
        Convert.ToBase64String(data).TrimEnd('=').Replace('+', '-').Replace('/', '_');

    private static byte[] Base64UrlDecode(string input)
    {
        var b64 = input.Replace('-', '+').Replace('_', '/');
        switch (b64.Length % 4)
        {
            case 2: b64 += "=="; break;
            case 3: b64 += "="; break;
        }
        return Convert.FromBase64String(b64);
    }
}

public sealed class SnapshotTokenException : Exception
{
    public SnapshotTokenException(string message) : base(message) { }
}

File: elixir/lib/envdoctor/runtime/token.ex
defmodule Envdoctor.Runtime.Token do
  @moduledoc """
  Portable snapshot token: `envd1:` + base64url(gzip(json)). Compact enough to
  paste into an issue or chat, self-contained, and schema-versioned so a
  decoder can refuse a token from a newer envdoctor instead of mis-diffing it.
  """

  alias Envdoctor.Json
  alias Envdoctor.Runtime.Capture

  @prefix "envd1:"

  @doc "Encode a snapshot into a single-line, paste-safe token."
  def encode(snapshot) do
    @prefix <> (snapshot |> Json.encode!() |> :zlib.gzip() |> Base.url_encode64(padding: false))
  end

  @doc "Decode a token back into a snapshot. Errors are clear and non-raising."
  def decode(token) do
    trimmed = String.trim(token)

    if String.starts_with?(trimmed, @prefix) do
      payload = String.slice(trimmed, byte_size(@prefix)..-1//1)

      with {:ok, gz} <- Base.url_decode64(payload, padding: false),
           {:ok, json} <- gunzip(gz),
           {:ok, raw} <- Json.decode(json) do
        raw |> normalize() |> assert_readable()
      else
        _ -> {:error, "Corrupt snapshot token: could not decode."}
      end
    else
      {:error, "Not an envdoctor snapshot token (missing envd1: prefix)."}
    end
  end

  @doc "Parse raw JSON (from a `--output` file) into a validated snapshot."
  def parse_json(text) do
    case Json.decode(text) do
      {:ok, raw} -> raw |> normalize() |> assert_readable()
      :error -> {:error, "Invalid snapshot JSON."}
    end
  end

  defp gunzip(gz) do
    {:ok, :zlib.gunzip(gz)}
  rescue
    _ -> :error
  end

  # Normalize decoded JSON (string keys) into the atom-keyed snapshot shape.
  defp normalize(%{} = raw) do
    os = Map.get(raw, "os", %{})

    tools =
      raw
      |> Map.get("tools", [])
      |> List.wrap()
      |> Enum.map(fn t when is_map(t) ->
        %{
          tool: Map.get(t, "tool", ""),
          version: Map.get(t, "version", ""),
          resolvedFrom: Map.get(t, "resolvedFrom", "")
        }
      end)

    globals =
      raw
      |> Map.get("globals", %{})
      |> then(fn
        g when is_map(g) ->
          Map.new(g, fn {eco, pkgs} ->
            {eco,
             pkgs
             |> List.wrap()
             |> Enum.map(fn p when is_map(p) ->
               %{name: Map.get(p, "name", ""), version: Map.get(p, "version", "")}
             end)}
          end)

        _ ->
          %{}
      end)

    %{
      schema: Map.get(raw, "schema"),
      capturedAt: Map.get(raw, "capturedAt", ""),
      os: %{
        platform: Map.get(os, "platform", ""),
        arch: Map.get(os, "arch", ""),
        release: Map.get(os, "release", "")
      },
      tools: tools,
      path: raw |> Map.get("path", []) |> List.wrap(),
      globals: globals,
      envFlagNames: raw |> Map.get("envFlagNames", []) |> List.wrap()
    }
  end

  defp assert_readable(%{schema: schema, tools: tools} = snapshot) do
    cond do
      not is_integer(schema) or not is_list(tools) ->
        {:error, "Not a runtime snapshot."}

      schema > Capture.snapshot_schema() ->
        {:error,
         "Snapshot schema v#{schema} is newer than this envdoctor (v#{Capture.snapshot_schema()}). Upgrade to compare it."}

      true ->
        {:ok, snapshot}
    end
  end
end
